### PR TITLE
resolves #21038

### LINKED ISSUE #21038: inconsistent use of assert
##### System information (version)
- OpenCV => 4.5.4
- Operating System / Platform => Linux
- Compiler => clang 12

##### Detailed description

There are 243 instances of assert in the code (not in tests nor 3rd_party) while CV_Assert should probably be used:

```
find ./modules -name 'test' -prune -o -type f -print | xargs grep " assert\( \)*(" | wc -l
```

Should indeed CV_Assert be used ? Or is there a reason for using assert (performance maybe).

This triggers invalid bugs in fuzzers that expect normal errors to be handled though a proper CV_Assert.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
 - [x] I updated to latest OpenCV version and the issue is still there

### PR DESCRIPTION
**Pull request by take1014**
resolves #21038
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by take1014** [comment]
@alalek  
I'm getting this warning  with compiling on linux opencl.
warning: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Wstrict-overflow]

on this line:
CV_Assert( (from_idx > to_idx && !elem) || (from_idx < to_idx && elem) );

I can't understand this warning correctly.
I think this warning was generated because it was determined to be unnecessary for optimization.
May I remove this assert on this line?

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/core/test/test_ds.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ static void cvTsSimpleSeqShiftAndCopy( CvTsSimpleSeq* seq, int from_idx, int to_
 
     if( from_idx == to_idx )
         return;
-    assert( (from_idx > to_idx && !elem) || (from_idx < to_idx && elem) );
+    CV_Assert( (from_idx > to_idx && !elem) || (from_idx < to_idx && elem) );
```
Comment (alalek) [inline modules/core/test/test_ds.cpp:53]: Please use this workaround:

```
    if (elem)
        CV_Assert(from_idx < to_idx);
    else
        CV_Assert(from_idx > to_idx);
```

Comment (take1014) [reply modules/core/test/test_ds.cpp]: @alalek 
Thanks for your review. I fixed them.

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Pull request merged**